tests/phpunit/tests/functions/moveDir.php

```diff
@@ -0,0 +1,281 @@
+		self::$existing_from_subdir      = self::$existing_from . 'existing_from_subdir/';
+		self::$existing_from_file        = self::$existing_from . 'existing_from_file.txt';
+		self::$existing_from_subdir_file = self::$existing_from_subdir . 'existing_from_subdir_file.txt';
+		self::$existing_to               = self::$test_dir . 'existing_to/';
+		self::$existing_to_file          = self::$existing_to . 'existing_to_file.txt';
+	}
+
+	/**
+	 * Sets up the directory structure before each test.
+	 */
+	public function set_up() {
+		global $wp_filesystem;
+
+		parent::set_up();
+
+		// Create the root directory.
+		$wp_filesystem->mkdir( self::$test_dir );
+
+		// Create the "from" directory structure.
+		$wp_filesystem->mkdir( self::$existing_from );
+		$wp_filesystem->touch( self::$existing_from_file );
+		$wp_filesystem->mkdir( self::$existing_from_subdir );
+		$wp_filesystem->touch( self::$existing_from_subdir_file );
+
+		// Create the "to" directory structure.
+		$wp_filesystem->mkdir( self::$existing_to );
+		$wp_filesystem->touch( self::$existing_to_file );
+	}
+
+	/**
+	 * Removes the test directory structure after each test.
+	 */
+	public function tear_down() {
+		global $wp_filesystem;
+
+		// Delete the root directory and its contents.
+		$wp_filesystem->delete( self::$test_dir, true );
+
+		parent::tear_down();
+	}
+
+	/**
+	 * Restores the filesystem after all tests have run.
+	 */
+	public static function tear_down_after_class() {
+		global $wp_filesystem;
+
+		// Restore the filesystem.
+		$wp_filesystem = self::$wp_filesystem_backup;
+
+		parent::tear_down_after_class();
+	}
+
+	/**
+	 * Tests that move_dir() returns a WP_Error object.
+	 *
+	 * @ticket 57375
+	 *
+	 * @dataProvider data_should_return_wp_error
+	 *
+	 * @param string $from      The source directory path.
+	 * @param string $to        The destination directory path.
+	 * @param bool   $overwrite Whether to overwrite the destination directory.
+	 * @param string $expected  The expected WP_Error code.
+	 */
+	public function test_should_return_wp_error( $from, $to, $overwrite, $expected ) {
+		global $wp_filesystem;
+
+		$from   = self::$test_dir . $from;
+		$to     = self::$test_dir . $to;
+		$result = move_dir( $from, $to, $overwrite );
+
+		$this->assertWPError(
+			$result,
+			'move_dir() did not return a WP_Error object.'
+		);
+
+		$this->assertSame(
+			$expected,
+			$result->get_error_code(),
+			'The expected error code was not returned.'
+		);
+
+		$this->assertTrue(
+			$wp_filesystem->exists( $from ),
+			'The $from directory does not exist anymore.'
+		);
+
+		if ( false === $overwrite && 'existing_to' === untrailingslashit( $to ) ) {
+			$this->assertTrue(
+				$wp_filesystem->exists( $to ),
+				'The $to directory does not exist anymore.'
+			);
+		}
+	}
+
+	/**
+	 * Data provider for test_should_return_wp_error().
+	 *
+	 * @return array[]
+	 */
+	public function data_should_return_wp_error() {
```

Comment:
Are you planning to add more items to this data provider? DPs of one are best avoided.

Reply:
At the moment, it's difficult to cover some of the later code due to how failures are produced: Such as changing the owner of a directory. The data provider is here at the moment because there's some improvements to be made to `WP_Filesystem_MockFS`, which should enable us to hit the other paths. In lieu of that, manual testing has proved successful.